docs/connections.md

```diff
@@ -0,0 +1,597 @@
+
+If the operator appears in the function name position, it can be defined as
+well:
+```
+«^» a n = a * a ^ (n - 1)
+```
+
+This introduces name `^` into the scope. It uses already defined `*` and `-`
+operators. (to avoid clutter the operators are not marked with »«)
+
+### Pattern matching
+If the assignment's left-hand side is a prefix application chain where the
+function is a constructor name, it will be desugared into a pattern match.
+
+Example:
+```
+»Some« «value» = »get_opt«
+```
+
+will be desugared into:
+
+```
+«value» = case »get_opt« of
+  »Some« b -> b
+  _      -> error 
+```
+
+Therefore, only `value` will be introduced into the parent scope. `Some` and
+`get_opt` must be defined, the former being an atom with at least single field.
+
+Using operators in the infix position will also attempt to pattern match its
+operands. For example:
+```
+«x»,«y» = »get_position«
+```
+
+TODO Describe what it gets desugared to when there is more than one variable
+name introduced? Now it is inconsistent, if we say that only canonical form of
+assignment is `variable_name = value`.
+
+### Extensions methods
+If the application target uses accessor operator `.`, e.g. `Int.add`, the last
+segment of target is the introduced identifier and the previous segments are
+used to type the implicit `this` parameter. 
+
+For example:
+```
+»Foo«.«bar» = 5
+```
+translated to:
+```
+«bar» this:»Foo« = 5
+```
+
+Which is then desugared into a lambda. The introduced name is only `bar`.
+
+If there are any prefix application arguments following the accessor-style
+target, they will be treated as arguments following implicit `this`.
+
+### Overloading
+Only the methods that take `this` as the first parameter can be overloaded. Each
+overload of the given name must have different type of `this`.
+
+However, the type of `this` will be often inferred by the typechecker and it
+IDE cannot tell if given overloads are valid or not.
+
+Example:
+
+```
+«foo» this:«a» = »body1«
+
+«foo» this:«b» = »body2«
+```
+
+In this case `a` and `b` for each `foo` definition will be inferred by the
+compiler. If they end up being different types, overloads are valid. If they are
+the same, an error will be raised.
+
+### Root scope assignments
+
+Any binding in the root scope gets an implicit `this` parameter that describes a
+module. Example:
+```
+main = print "Hello"
+```
+
+`main` here is a function binding that is desugared to lambda. As such its body
+has its own, separate scope. That would not have been a case, if such lines
+appeared in any non-root code block.
+
+## Lambdas
+`arg -> value` is the syntax for lambdas. Left-hand side is a pattern for the
+argument (lambdas are always unary) and the right-hand side is its body.
+Right-hand side, i.e. the lambda body, introduces its own scope.
+
+If the lambda is *not* introduced in what is already a pattern context, the `->`'s pattern introduces identifiers into the scope of the right-hand
```

Comment:
I think you introduced _pattern scopes_, and _pattern context_ is a term from some previous version of this doc.